compute/autoscaler/demo/frontend.py

```diff
@@ -0,0 +1,107 @@
+# limitations under the License.
+#
+"""A simple app which listens to HTTP GET requests and consumes CPU in response.
+
+This binary runs in a GCE VM. It serves HTTP requests on port 80. Every request
+with path '/service' consumes 1 core-second of CPU time, with the timeout of
+5 (walltime) seconds. The purpose of this application is to demonstrate how
+Google Compute Engine Autoscaler can scale a web frontend server based on CPU
+utilization.
+
+The original version of this file is available here:
+https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/compute/
+    autoscaler/demo/tests/test_frontend.py
+"""
+
+import BaseHTTPServer
+from multiprocessing import Process
+import os
+import SocketServer
+import sys
+import thread
+import time
+
+
+class CpuBurner(object):
+  REQUEST_CPUTIME_SEC = 1.0
+  REQUEST_TIMEOUT_SEC = 5.0
+
+  def GetWallTime(self):
+    return time.time()
+
+  def GetUserCpuTime(self):
+    return os.times()[0]
+
+  def BusyWait(self):
+    for _ in xrange(100000):
+        pass
+
+  def BurnCpu(self):
+    """Consume REQUEST_CPUTIME_SEC core seconds.
+
+    This method consumes REQUEST_CPUTIME_SEC core seconds. If unable to complete
+    within REQUEST_TIMEOUT_SEC walltime seconds, it times out and terminates the
+    process."""
+    start_walltime_sec = self.GetWallTime()
+    start_cputime_sec = self.GetUserCpuTime()
+    while self.GetUserCpuTime() < start_cputime_sec + CpuBurner.REQUEST_CPUTIME_SEC:
+      self.BusyWait()
+      if self.GetWallTime() > start_walltime_sec + CpuBurner.REQUEST_TIMEOUT_SEC:
+        sys.exit(1)
+
+  def HandleHttpRequest(self):
+    """Process a request to consume CPU and produce an HTTP response."""
+    start_time = self.GetWallTime()
+    p = Process(target=self.BurnCpu)  # Run in a separate process.
+    p.start()
+    # Force kill after timeout + 1 sec.
+    p.join(timeout=CpuBurner.REQUEST_TIMEOUT_SEC+1)
+    if p.is_alive():
+      p.terminate()
+    if p.exitcode != 0:
+      return (500, "Request failed\n")
+    else:
+      end_time = self.GetWallTime()
+      response = "Request took %.2f walltime seconds\n" % (end_time - start_time)
+      return (200, response)
+
+
+class DemoRequestHandler(BaseHTTPServer.BaseHTTPRequestHandler):
+  """Request handler for Demo http server."""
+
+  def do_GET(self):
+    """Handle an HTTP GET request."""
+    mapping = {
+        "/": lambda: (200, "OK"),  # Return HTTP 200 response.
+        "/service": CpuBurner().HandleHttpRequest(),
+    }
+    if self.path not in mapping:
+      self.send_response(404)
+      self.end_headers()
+      return
+    (code, response) = mapping[self.path]()
+    self.send_response(code)
+    self.end_headers()
+    self.wfile.write(response)
+    self.wfile.close()
+
+
+class DemoHttpServer(SocketServer.ThreadingMixIn,
+                     BaseHTTPServer.HTTPServer):
+  pass
+
+
+if __name__ == "__main__":
+  httpd = DemoHttpServer(("", 80), DemoRequestHandler)
```

Comment:
The instructions we have for deploying the instance is basically to set the following startup script: "wget https://storage.googleapis.com/autoscaler-demo/frontend.py; sudo python frontend&" (see slide 23 of the video script: https://docs.google.com/presentation/d/1BfKmGQAuV7Qq0vj4wHntrHcZxECoaJaEHTuW5COdMFM/edit#slide=id.gb796329e8_1_0).

I have to confess that we've already done the video recording and we're at the post-processing stage, which gives us little flexibility with the actual setup of the demo. We're still free to change the code sample, but we can't add additional steps such as using custom images.

Our goal was to keep the demo as short and self contained as possible to convince users that the basic setup can be done in a very few steps. You're totally right there is a trade-off between the simplicity of the code and the setup.

As for the code: I'd love to use something higher level than BaseHTTPServer if it lived in a standard module..